pydantic/config.py

```diff
@@ -1017,9 +1019,17 @@ class Model(TypedDict):
         ```
     """
 
-    def inner(TypedDictClass: _TypeT, /) -> _TypeT:
-        TypedDictClass.__pydantic_config__ = config
-        return TypedDictClass
+    def inner(class_: _TypeT, /) -> _TypeT:
+        # Ideally, we would check for `class_` to either be a `TypedDict` or a stdlib dataclass.
+        # However, the `@with_config` decorator can be applied *after* `@dataclass`. To avoid
+        # common mistakes, we at least check for `class_` to not be a Pydantic model.
+        if is_model_class(class_):
```

Comment:
@Viicos,

My initial hunch is that you can do the `is_model_class` import in this function, and maybe that will fix the failing test?